modules/module-postgres/src/replication/WalStream.ts

```diff
@@ -342,25 +342,35 @@ WHERE  oid = $1::regclass`,
     await this.storage.startBatch(
       { zeroLSN: ZERO_LSN, defaultSchema: POSTGRES_DEFAULT_SCHEMA, storeCurrentData: true, skipExistingRows: true },
       async (batch) => {
-        const rs = await db.query(`select pg_current_wal_lsn() as lsn`);
-        const startLsn = rs.rows[0][0];
-
         for (let tablePattern of sourceTables) {
           const tables = await this.getQualifiedTableNames(batch, db, tablePattern);
           for (let table of tables) {
             if (table.snapshotComplete) {
               logger.info(`${this.slot_name} Skipping ${table.qualifiedName} - snapshot already done`);
               continue;
             }
-            await this.snapshotTable(batch, db, table);
+            let tableLsnNotBefore: string;
+            await db.query('BEGIN');
+            try {
+              await this.snapshotTable(batch, db, table);
+
+              const rs = await db.query(`select pg_current_wal_lsn() as lsn`);
+              tableLsnNotBefore = rs.rows[0][0];
+            } finally {
+              // Read-only transaction, commit does not actually do anything.
+              await db.query('COMMIT');
+            }
 
-            const rs = await db.query(`select pg_current_wal_lsn() as lsn`);
-            const tableLsnNotBefore = rs.rows[0][0];
             await batch.markSnapshotDone([table], tableLsnNotBefore);
             await touch();
           }
         }
-        await batch.commit(startLsn);
+
+        // Always commit the initial snapshot at zero.
+        // This makes sure we don't skip any changes applied before starting this snapshot,
+        // in the case of snapshot retries.
+        // We could alternatively commit at the replication slot LSN.
+        await batch.commit(ZERO_LSN);
```

Comment:
This was not an issue before. However, changing the snapshot to execute in a transaction, introduced consistency issues - streaming replication after the snapshot could skip changes that shouldn't be skipped.

I'm not sure why changing between using/not using a transaction has an effect on this, but the real issue was doing a commit with the wrong LSN.

There is a small chance that there are more scenarios triggering the same consistency issue since #150 (only when initial replication is interrupted and retried).